core/src/driver/scheduler/evm.rs

```diff
@@ -32,110 +37,118 @@ impl EvmScheduler {
-            Ok(solution) => {
-                info!("successfully solved batch {}", batch_id);
-                self.last_batch = Some(batch_id);
-                // TODO: handle earliest_solution_submit_time .
-                match self
-                    .driver
-                    .submit_solution(batch_id.into(), solution)
-                    .wait()
-                {
-                    Ok(()) => info!("successfully submitted solution for batch {}", batch_id),
-                    Err(err) => error!(
-                        "failed to submit solution for batch {}: {:?}",
-                        batch_id, err
-                    ),
+    async fn solve(&self, batch_id: u32) -> Result<Option<Solution>> {
+        while let Some(time_limit) = self.solver_time_limit(batch_id).await? {
+            info!(
+                "solving for batch {} with time limit {}s",
+                batch_id,
+                time_limit.as_secs_f64(),
+            );
+            match self.driver.solve_batch(batch_id.into(), time_limit).await {
+                Ok(solution) => {
+                    info!("successfully solved batch {}", batch_id);
+                    return Ok(Some(solution));
+                }
+                Err(DriverError::Retry(err)) => {
+                    error!("driver retryable error for batch {}: {:?}", batch_id, err);
+                }
+                Err(DriverError::Skip(err)) => {
+                    error!("driver error for batch {}: {:?}", batch_id, err);
+                    return Ok(None);
                 }
-            }
-            Err(DriverError::Retry(err)) => {
-                error!("driver retryable error for batch {}: {:?}", batch_id, err);
-            }
-            Err(DriverError::Skip(err)) => {
-                error!("driver error for batch {}: {:?}", batch_id, err);
-                self.last_batch = Some(batch_id);
             }
         }
+        warn!("skipping batch {} because we ran out of time", batch_id);
+        Ok(None)
+    }
 
-        Ok(())
+    async fn submit(&self, batch_id: u32, solution: Solution) {
+        // TODO: handle earliest_solution_submit_time .
+        match self.driver.submit_solution(batch_id.into(), solution).await {
+            Ok(()) => info!("successfully submitted solution for batch {}", batch_id),
+            Err(err) => error!(
+                "failed to submit solution for batch {}: {:?}",
+                batch_id, err
+            ),
+        }
     }
 
-    /// Gets the current solving batch ID.
-    ///
-    /// This is the current batch ID minus 1, as the current batch ID is the ID
-    /// of the batch that is currently accepting orders.
-    fn current_solving_batch(&self) -> Result<u32> {
-        Ok(self.exchange.get_current_auction_index().wait()? - 1)
+    /// Wait for and solve the next batch.
+    async fn step(&self, last_batch: Option<u32>) -> Result<u32> {
+        let last_batch = match last_batch {
```

Comment:
Is it possible to skip a batch if the previous run takes longer than the batch duration?

**edit**: NVM, I see the previous batch is stored in the run loop.